rfcs/0000-frontier-listener.md

```diff
@@ -0,0 +1,76 @@
+## Summary
+[summary]: #summary
+
+Allow properties based the transition frontier to be efficiently tracked.
+
+## Motivation
+
+[motivation]: #motivation
+
+There are several properties that we might want to obtain from the data in the transition frontier that
+are too expensive to calculate by iterating over all its transitions. However, we don't want to add complexity
+to the transition frontier by having it keep track of these values that are not strictly necessary for its
+operation and are only used externally.
+
+As an example, the snark pool never has elements removed and requires some system of garbage collection.
+One way to do this would be to track a reference count for each piece of `Work` in the table of proofs, but this
+table of references probably shouldn't be part the transition frontier itself as it's completely unrelated.
+
+We'd like to have a way for other parts of the system to incrementally build up datastructures based on the
+breadcrumbs that are being added and removed to the frontier.
+
+Other examples of such an abstraction:
+  - managing the transaction pool
+  - tracking information for consensus optimization
+
+## Detailed design
+
+[detailed-design]: #detailed-design
+
+This design is based on the introduction of a new signature in `Protocols.coda_transition_frontier`:
+```ocaml
+module type Transition_frontier_listener_intf = sig
+  type t
+  val add_breadcrumb : t -> Breadcrumb.t -> unit;
+  val remove_breadcrumb : t -> Breadcrumb.t -> unit;
+end
```

Comment:
I find hiding information in closures something that makes code harder to read and trace -- so I prefer the first way.

Reply:
I've eventually managed to prototype something out with first-class modules but I would be fairly on board with switching to closures as it seems like it would simplify the gymnastics involved in implementing/reading this code..